### PR TITLE
Fix Reading the Currency Formats

### LINKED ISSUE #946: System.Globalization.Native shim: Unknown Numeric Patterns
System.Globalization.Native shim had this assert (since forever): https://github.com/dotnet/runtime/blob/1983518fa4c901d35bb3ac7b1f76d7fa810c3543/src/coreclr/src/corefx/System.Globalization.Native/pal_localeNumberData.c#L203

This codepath was not exercised on debug builds of the shim due to test hole. I have confirmed it by test-only PR #890.  This test hole got fixed as side-effect of moving the shim under libraries (#875).

I am commenting our the assert against this issue. This issues is about investigating whether the unknown number patterns are expected or whether there are fixes required to recognize them.

### PR DESCRIPTION
**Pull request by tarekgh**
Fixes #946

We have encountered 2 locales that have a negative currency format which we failed to read them correctly. The fix here is to allow such formats and read them successfully. Also returning the removed assert which is useful in catching such cases.

```
luy       :  ¤- #,##0      // .NET format should be "C- n"
luy_KE    : ¤- #,##0      // .NET format should be "C- n"
```

### CODE REVIEW

**Comment by tarekgh** [comment]
CC @jkotas

File: src/libraries/Native/Unix/System.Globalization.Native/pal_localeNumberData.c [hunk 1]
```diff
@@ -220,6 +215,7 @@ static int GetCurrencyNegativePattern(const char* locale)
     static const char* Patterns[] = {"(Cn)",
                                      "-Cn",
                                      "C-n",
+                                     "C- n",
```
Comment (jkotas) [inline src/libraries/Native/Unix/System.Globalization.Native/pal_localeNumberData.c]: Is this going to shift indices of the remaining patterns, and in turn change the value returned by https://docs.microsoft.com/en-us/dotnet/api/system.globalization.numberformatinfo.currencynegativepattern?view=netframework-4.8 ?

Comment (tarekgh) [reply src/libraries/Native/Unix/System.Globalization.Native/pal_localeNumberData.c]: You are right. missed fixing the managed side too and should insert the value at the end of the table.

File: src/libraries/Common/src/System/Globalization/FormatProvider.Number.cs [hunk 2]
```diff
@@ -607,7 +607,7 @@ internal static unsafe bool TryStringToNumber(ReadOnlySpan<char> str, NumberStyl
                 "($#)", "-$#", "$-#", "$#-",
                 "(#$)", "-#$", "#-$", "#$-",
                 "-# $", "-$ #", "# $-", "$ #-",
-                "$ -#", "#- $", "($ #)", "(# $)"
+                "$ -#", "#- $", "($ #)", "(# $)", "$- #"
```
Comment (jkotas) [inline src/libraries/Common/src/System/Globalization/FormatProvider.Number.cs]: Nit: All lines above are 4 strings per line. Follow the style a put it on a new line?

**Comment by jkotas** [comment]
This may also need a doc update. The docs say that the max format number is 15.

**Comment by tarekgh** [comment]
I logged the doc issue https://github.com/dotnet/docs/issues/17464

File: src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs [hunk 3]
```diff
@@ -277,7 +277,7 @@ internal static partial class Number
             "($#)", "-$#", "$-#", "$#-",
             "(#$)", "-#$", "#-$", "#$-",
             "-# $", "-$ #", "# $-", "$ #-",
-            "$ -#", "#- $", "($ #)", "(# $)"
+            "$ -#", "#- $", "($ #)", "(# $)", "$- #"
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs]: Same here

**Review by jkotas — APPROVED** [review]

File: src/libraries/System.Globalization/tests/NumberFormatInfo/NumberFormatInfoTests.cs [hunk 4]
```diff
@@ -87,6 +87,20 @@ public void NativeDigitsTest()
...
+        public void TestNFIFormatLimits()
+        {
+            foreach (CultureInfo ci in CultureInfo.GetCultures(CultureTypes.AllCultures))
+            {
+                NumberFormatInfo nfi = ci.NumberFormat;
+                Assert.True(nfi.CurrencyNegativePattern <= 16);
+                Assert.True(nfi.CurrencyPositivePattern <= 3);
+                Assert.True(nfi.PercentNegativePattern <= 11);
+                Assert.True(nfi.PercentPositivePattern <= 3);
+                Assert.True(nfi.NumberNegativePattern <= 4);
```
Comment (stephentoub) [inline src/libraries/System.Globalization/tests/NumberFormatInfo/NumberFormatInfoTests.cs]: These can all use Assert.InRange.  The benefit of that is, if it fails, the assert will output the invalid value that was tested, whereas with Assert.True it'll just output something like "Expected: True, Actual: False".

**Review by stephentoub — APPROVED** [review]

**Comment by tarekgh** [comment]
Looks fedora having some other issue as the assert start to fail again. I'll take a look to see what other formats there is missing.

**Comment by tarekgh** [comment]
In Fedora, the en_US_POSIX and bg_BG locale formats not included the # character :

```
en_US_POSIX    :  '¤ 0.', and  '0%'
bg_BG               :  '0.00'
```

**Comment by tarekgh** [comment]
@jkotas @stephentoub I added a small delta to the change in the last commit to address the Fedora issue I mentioned in https://github.com/dotnet/runtime/pull/33599#issuecomment-599822470. I just wanted to let you know if I merge this PR. Thanks.

**Review by jkotas — APPROVED** [review]

**Pull request merged**